### PR TITLE
[block] Use the async file engine in unit tests on kernel >= 5.10

### PR DESCRIPTION
**Pull request by serban300**
# Reason for This PR

[block] Use the async file engine in unit tests on kernel >= 5.10

## Description of Changes

[block] Use the async file engine in unit tests on kernel >= 5.10

- [ ] This functionality can be added in [`rust-vmm`](https://github.com/rust-vmm).

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria.]`
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [x] All commits in this PR are signed (`git commit -s`).
- [x] The issue which led to this PR has a clear conclusion.
- [x] This PR follows the solution outlined in the related issue.
- [x] The description of changes is clear and encompassing.
- [x] Any required documentation changes (code and docs) are included in this PR.
- [x] Any newly added `unsafe` code is properly documented.
- [x] Any API changes are reflected in `firecracker/swagger.yaml`.
- [x] Any user-facing changes are mentioned in `CHANGELOG.md`.
- [x] All added/changed functionality is tested.

### CODE REVIEW

File: src/io_uring/src/lib.rs [hunk 1]
```diff
@@ -90,24 +98,48 @@ impl IoUring {
             len if fd < 0 || (len as i32 - 1) < fd => {
                 Err((Error::InvalidFixedFd(fd), op.user_data()))
             }
-            _ => self.squeue.push(unsafe { op.into_sqe() }).map_err(
-                |err_tuple: (SQueueError, T)| -> (Error, T) {
+            _ => self
+                .squeue
+                .push(unsafe { op.into_sqe() })
+                .map(|_| {
+                    self.to_submit = self.to_submit.saturating_add(1);
```
Comment (georgepisaltu) [inline src/io_uring/src/lib.rs:105]: How likely is is that we will get to saturating this value? Should we treat this in a separate error case or is it actually impossible to get to `u32::MAX` here?

File: src/io_uring/src/lib.rs [hunk 2]
```diff
@@ -90,24 +98,48 @@ impl IoUring {
...
+            .submit(self.to_submit, 0)
+            .map(|submitted| self.process_submitted(submitted))
+            .map_err(Error::SQueue)
     }
 
-    pub fn submit_and_wait(&mut self, min_complete: u32) -> Result<u32> {
-        self.squeue.submit(min_complete).map_err(Error::SQueue)
+    pub fn submit_and_wait_all(&mut self) -> Result<u32> {
+        self.squeue
+            .submit(self.to_submit, self.to_submit + self.to_pop)
```
Comment (georgepisaltu) [inline src/io_uring/src/lib.rs:140]: Why not check for overflow here too?

**Review by georgepisaltu — COMMENTED** [review]
Looks good overall!

File: src/devices/src/virtio/block/io/async_io.rs [hunk 3]
```diff
@@ -173,9 +173,9 @@ impl<T> AsyncFileEngine<T> {
...
     #[cfg(test)]
     // Useful when testing, to skip the eventfd polling and waiting.
-    pub fn kick_submission_queue_and_wait(&mut self, min_complete: u32) -> Result<(), Error> {
+    pub fn kick_submission_queue_and_wait_all(&mut self) -> Result<(), Error> {
         self.ring
-            .submit_and_wait(min_complete)
+            .submit_and_wait_all()
             .map(|_| ())
             .map_err(Error::IoUring)
     }
```
Comment (luminitavoicu) [inline src/devices/src/virtio/block/io/async_io.rs:181]: I think we should move all test-related functions at the end of the file, for improved visibility.

File: src/devices/src/virtio/block/io/async_io.rs [hunk 4]
```diff
@@ -197,19 +197,10 @@ impl<T> AsyncFileEngine<T> {
...
-            }
-
-            self.ring.submit_and_wait(sq_len).map_err(Error::IoUring)?;
-        }
-
-        Ok(())
+        self.ring
+            .submit_and_wait_all()
+            .map(|_| ())
+            .map_err(Error::IoUring)
```
Comment (luminitavoicu) [inline src/devices/src/virtio/block/io/async_io.rs:203]: This function has the same body as `kick_submission_queue_and_wait_all` above, which is used for testing purposes. Why do we need both? Am I missing something?

File: src/io_uring/src/lib.rs [hunk 5]
```diff
@@ -46,6 +46,12 @@ pub struct IoUring {
     registered_fds_count: u32,
     squeue: SubmissionQueue,
     cqueue: CompletionQueue,
+
+    // Number of ops yet to be submitted.
+    to_submit: u32,
+    // Number of ops yet to be pop-ed from the CQ. These ops either haven't been pop-ed yet,
```
Comment (luminitavoicu) [inline src/io_uring/src/lib.rs:52]: This line seems too long.

Comment (serban300) [reply src/io_uring/src/lib.rs:50]: Not sure. I thought comments should be under 100 characters.

File: src/io_uring/src/lib.rs (hunk 1 above)
Comment (serban300) [reply src/io_uring/src/lib.rs]: it's impossible to get u32::MAX here. I added comments.

File: src/io_uring/src/lib.rs (hunk 2 above)
Comment (serban300) [reply src/io_uring/src/lib.rs]: We shouldn't get an overflow here since `self.to_submit < 32768` and `self.to_pop < 32768 * 2`. I added comments.

**Review by alindima — COMMENTED** [review]
Looks good! I only have a nit

**Review by luminitavoicu — APPROVED** [review]
LGTM

File: src/io_uring/src/lib.rs (hunk 5 above)
Comment (luminitavoicu) [reply src/io_uring/src/lib.rs:50]: You are right, this is fine!

File: src/devices/src/virtio/block/test_utils.rs [hunk 6]
```diff
@@ -49,14 +56,47 @@ pub fn rate_limiter(blk: &mut Block) -> &RateLimiter {
     &blk.rate_limiter
 }
 
+#[cfg(test)]
+fn assert_pending_irq(b: &Block, maybe_expected_irq: Option<bool>) {
```
[the remark this answers is not in the document]
Comment (serban300) [reply src/devices/src/virtio/block/test_utils.rs]: Done. I removed this method.

**Review by alindima — DISMISSED** [review]

**Review by alindima — APPROVED** [review]

**Review by luminitavoicu — APPROVED** [review]

**Pull request merged**